src/sys/petsc_interface.cxx

```diff
@@ -0,0 +1,173 @@
+IndexerPtr GlobalIndexer::getInstance(Mesh* localmesh) {
+  // Check that the identity of bout::globals::mesh hasn't changed
+  // since last call. (Needed for unit tests)
+  if (bout::globals::mesh != globalmesh) {
+    globalmesh = bout::globals::mesh;
+    initialisedGlobal = false;
+  }
+  if (localmesh == globalmesh) {
+    if (!initialisedGlobal) {
+      globalInstance = std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+      initialisedGlobal = true;
+    }
+    return globalInstance;
+  } else {
+    return std::shared_ptr<GlobalIndexer>(new GlobalIndexer(localmesh));
+  }
+}
+
+void GlobalIndexer::initialiseTest() {
+  registerFieldForTest(indices3D);
+  registerFieldForTest(indices2D);
+  registerFieldForTest(indicesPerp);
+}
+
+void GlobalIndexer::initialise() {
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+  // Communicate a second time to get any corner values
+  fieldmesh->communicate(indices3D, indices2D);
+  fieldmesh->communicate(indicesPerp);
+}
+
+Mesh* GlobalIndexer::getMesh() { return fieldmesh; }
+
+PetscInt GlobalIndexer::getGlobal(Ind2D ind) {
+  return static_cast<PetscInt>(indices2D[ind] + 0.5);
+}
+
+PetscInt GlobalIndexer::getGlobal(Ind3D ind) {
+  return static_cast<PetscInt>(indices3D[ind] + 0.5);
+}
+
+PetscInt GlobalIndexer::getGlobal(IndPerp ind) {
+  return static_cast<PetscInt>(indicesPerp[ind] + 0.5);
+}
+
+void GlobalIndexer::registerFieldForTest(FieldData& UNUSED(f)) {
+  // This is a place-holder which does nothing. It can be overridden
+  // by descendent classes if necessary to set up testing.
+  return;
+}
+
+void GlobalIndexer::registerFieldForTest(FieldPerp& UNUSED(f)) {
+  // This is a place-holder which does nothing. It can be overridden
+  // by descendent classes if necessary to set up testing.
+  return;
+}
+
+GlobalIndexer::GlobalIndexer(Mesh* localmesh)
+    : fieldmesh(localmesh), indices3D(-2., localmesh), indices2D(-2., localmesh),
+      indicesPerp(-2., localmesh) {
+  // Set up the 3D indices
+  int counter = localmesh->globalStartIndex3D();
+  for (RangeIterator it = localmesh->iterateBndryLowerY(); !it.isDone(); it++) {
+    for (int z = 0; z < localmesh->LocalNz; z++) {
+      if (it.ind == localmesh->xstart)
+        indices3D(it.ind - 1, localmesh->ystart - 1, z) = counter++;
+      if (it.ind == localmesh->xend)
+        indices3D(it.ind + 1, localmesh->ystart - 1, z) = counter++;
+      indices3D(it.ind, localmesh->ystart - 1, z) = counter++;
+    }
+  }
+  for (RangeIterator it = localmesh->iterateBndryUpperY(); !it.isDone(); it++) {
+    for (int z = 0; z < localmesh->LocalNz; z++) {
+      if (it.ind == localmesh->xstart)
+        indices3D(it.ind - 1, localmesh->yend + 1, z) = counter++;
+      if (it.ind == localmesh->xend)
+        indices3D(it.ind + 1, localmesh->yend + 1, z) = counter++;
+      indices3D(it.ind, localmesh->yend + 1, z) = counter++;
+    }
+  }
+  BOUT_FOR_SERIAL(i, localmesh->getRegion3D("RGN_NOY")) {
```

Comment:
I'm thinking I might do this now, actually, as I find myself needing to iterate over boundary regions on a number of different occasions. The existing `RGN_INNER_X`, `RGN_LOWER_Y`, etc. almost do the job, except they include more than one cell in the x-direction and they don't include corner cells. It would be quite easy to create some additional regions for this purpose in `BoutMesh::addBoundaryRegions()`, although I'm not quite sure what to call them.

Reply:
I'm not sure we've got a good term for "the first point in the boundary". Maybe even just `RGN_LOWER_Y_FIRST_POINT_ONLY` as a placeholder for now?

I suspect we will need to come up with some good terminology soon, as those regions will also be needed for boundary conditions where we want each point in the plane of the boundary, plus some arbitrary number of points perpendicular to it.

Other suggestions:

- `RGN_*_PLANE`
- `RGN_*_SLICE`
- `RGN_*_PRIMARY`
- `RGN_*_HEAD`
- `RGN_*_FRONT`
- `RGN_*_MINOR`
- `RGN_*_BEGIN`

`begin` is quite appealing -- mirrors the names for iterators. Of course, it does imply `RGN_*_END` which is probably not as useful